File: Backend/zurinet-server/googleMaps/extractLocations.py
import sys
import json
import spacy

# Load spaCy English model
nlp = spacy.load("en_core_web_sm")

def extract_locations(text):
    doc = nlp(text)
    return [ent.text for ent in doc.ents if ent.label_ in ["GPE", "LOC"]]

if __name__ == "__main__":
    # Get JSON from Node.js stdin
    data = json.loads(sys.stdin.read())
    articles = data.get("articles", [])
    
    locations = []
    for article in articles:
        text = f"{article.get('title', '')} {article.get('description', '')} {article.get('content', '')}"
        locations.extend(extract_locations(text))
    
    print(json.dumps(list(set(locations))))  # Return unique locations

File: Backend/zurinet-server/extractLocations.js
import nlp from "compromise";

/**
 * Extract locations (places) from text using compromise
 * @param {string} text
 * @returns {string[]} list of unique location names
 */
export function extractLocations(text) {
  const doc = nlp(text);
  const places = doc.places().out("array"); // extract place names
  return [...new Set(places)]; // unique values
}

/**
 * Extract locations from an array of articles
 * @param {Array} articles - news articles with title, description, content
 * @returns {string[]} unique location names from all articles
 */
export function extractLocationsFromArticles(articles) {
  let allLocations = [];

  for (const article of articles) {
    const text = [article.title, article.description, article.content]
      .filter(Boolean)
      .join(" ");
    allLocations = allLocations.concat(extractLocations(text));
  }

  return [...new Set(allLocations)];
}
